cirq/ops/pauli_string.py

```diff
@@ -240,6 +239,187 @@ def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs):
+            cirq.X(q0).expectation(state, qubit_map={q0: 1, q1: 0}) = 1
+
+        Args:
+            state: An array representing a valid wavefunction.
+            qubit_map: A map from all qubits used in this PauliString to the
+            indices of the qubits that `state` is defined over.
+
+        Returns:
+            The expectation value of the input state.
+
+        Raises:
+            NotImplementedError if this PauliString is non-Hermitian.
+        """
+        if abs(self.coefficient.imag) > 0.0001:
+            raise NotImplementedError(
+                "Cannot compute expectation value of a non-Hermitian "
+                "PauliString <{}>. Coefficient must be real.".format(self))
+
+        # FIXME: Avoid enforce specific complex type. This is necessary to
+        # prevent an `apply_unitary` bug (Issue #2041).
+        if state.dtype.kind != 'c':
+            raise TypeError("Input state dtype must be np.complex64 or "
+                            "np.complex128")
+
+        size = state.size
+        num_qubits = size.bit_length() - 1
+        if len(state.shape) != 1 and state.shape != (2,) * num_qubits:
+            raise ValueError("Input array does not represent a wavefunction "
+                             "with shape `(2 ** n,)` or `(2, ..., 2)`.")
+
+        _validate_qubit_mapping(qubit_map, self.qubits, num_qubits)
+        # HACK: avoid circular import
+        from cirq.sim.wave_function import validate_normalized_state
+        validate_normalized_state(state, num_qubits, dtype=state.dtype)
+        return self._expectation_from_wavefunction_no_validation(state, qubit_map)
+
+    def _expectation_from_wavefunction_no_validation(self,
+                                       state: np.ndarray,
+                                       qubit_map: Mapping[raw_types.Qid, int]
+                                       ) -> float:
+        """Evaluate the expectation of this PauliString given a wavefunction.
+
+        This method does not provide input validation. See
+        `PauliString.expectation_from_wavefunction` for function description.
+
+        Args:
+            state: An array representing a valid wavefunction.
+            qubit_map: A map from all qubits used in this PauliString to the
+            indices of the qubits that `state` is defined over.
+
+        Returns:
+            The expectation value of the input state.
+        """
+        if len(state.shape) == 1:
+            num_qubits = state.shape[0].bit_length() - 1
+            state = np.reshape(state, (2,) * num_qubits)
+
+        ket = np.copy(state)
+        for qubit, pauli in self.items():
+            buffer = np.empty(ket.shape, dtype=state.dtype)
+            args = protocols.ApplyUnitaryArgs(target_tensor=ket,
+                                              available_buffer=buffer,
+                                              axes=(qubit_map[qubit],))
+            ket = protocols.apply_unitary(pauli, args)
```

Comment:
I like how you did this - O(n) expectation values without writing any new code.

Reply:
what if we want to calculate the expectation value of something hermetian but non-unitary (For example a projector on the zero state of a qubit)?

Reply:
Credit for that code actually goes to Matt.

Non-unitary is fine (PauliSums will be non-unitary on general)

Reply:
i suppose that can be expressed as (I-Z)/2. I guess anything expressible as a pauli_string will have an expectation value calculable via unitary evolution - correct me if i'm wrong though.

Reply:
yep. For these methods, the expectation is purely postprocessing on a circuit output.